### PR TITLE
feat(e2e): e2e tests for comments

### PR DESCRIPTION
**Pull request by caando**
Added E2E tests for comments feature of review requests.

Closes IS-258

**Note**

Occasionally `cy.request` fails and return undefined for reasons unknown. This will cause the tests to break.

## Tests

- [ ] Unit tests (using `npm run tests`)
- [x] e2e tests (comment on this PR with the text `!run e2e`)
- [ ] Smoke tests

## Deploy Notes

Requires #1338 to be merged first.

### CODE REVIEW

File: cypress/api/reviewRequest.api.ts [hunk 1]
```diff
@@ -30,11 +30,25 @@ export const listReviewRequests = (): Cypress.Chainable<{ id: number }[]> => {
...
-  cy.request("POST", `${BASE_URL}/${id}/approve`)
+export const approveReviewRequest = (
+  id: number
+): Cypress.Chainable<Cypress.Response<any>> => {
+  return cy.request("POST", `${BASE_URL}/${id}/approve`)
+}
+
+export const mergeReviewRequest = (
+  id: number
+): Cypress.Chainable<Cypress.Response<any>> => {
```
Comment (seaerchin) [inline cypress/api/reviewRequest.api.ts:46]: for the below 2, can we give the type as `void` instead of `any`?

File: cypress/e2e/comments.spec.ts [hunk 2]
```diff
@@ -28,17 +30,33 @@ const MOCK_COMMENT = "mock comment"
...
+    .contains(userEmail)
+    .should("be.visible")
+}
+
+const checkCommentDisabled = () =>
+  cy.get(COMMENTS_DRAWER_BUTTON_SELECTOR).should("be.disabled")
+
 const USER_NOT_AUTHORISED_COMMENT_ERROR_MESSAGE =
   "Only collaborators of a site can view review request comments!"
+const NOTIFICATION_CANNOT_BE_RETRIVED_ERROR_MESSAGE =
```
Comment (seaerchin) [inline cypress/e2e/comments.spec.ts:50]: these seem to be unused so feel free to remove em

File: cypress/e2e/comments.spec.ts [hunk 3]
```diff
@@ -100,32 +114,73 @@ describe("Comments", () => {
...
     it("should not allow empty comments to be posted", () => {
       // Arrange
+      visitE2eEmailTestRepo()
+
       // Act
+      openReviewRequest()
+      cy.wait(`@${COMMENT_INTERCEPTOR}`)
+      ignoreResizeError()
+      openCommentsDrawer()
+      getCommentInput().type(MOCK_COMMENT).clear().blur()
```
Comment (seaerchin) [inline cypress/e2e/comments.spec.ts:129]: can we also check that it is disabled on initial load?

File: cypress/e2e/comments.spec.ts [hunk 4]
```diff
@@ -28,17 +30,33 @@ const MOCK_COMMENT = "mock comment"
...
 
-const submitComment = () => cy.get(SUBMIT_COMMENT_BUTTON_SELECTOR).click()
 const COMMENT_INTERCEPTOR = "getComments"
 
 describe("Comments", () => {
   beforeEach(() => {
     cy.setEmailSessionDefaults("Email admin")
+    cy.intercept("GET", "**/comments").as(COMMENT_INTERCEPTOR)
     cy.setupDefaultInterceptors()
-    visitE2eEmailTestRepo()
```
Comment (seaerchin) [inline cypress/e2e/comments.spec.ts:41]: why was this line deleted?

File: cypress/e2e/comments.spec.ts [hunk 5]
```diff
@@ -175,46 +228,129 @@ describe("Comments", () => {
...
         // before the redirect is done
-        openCommentsDrawer()
-        cy.get(COMMENTS_INPUT_SELECTOR).type(MOCK_COMMENT).blur()
-        submitComment()
 
-        // Assert
-        cy.contains(USER_NOT_AUTHORISED_COMMENT_ERROR_MESSAGE).should(
-          "be.visible"
-        )
+        // // Assert
```
Comment (seaerchin) [inline cypress/e2e/comments.spec.ts:232]: nit: remove double comment

File: cypress/e2e/comments.spec.ts [hunk 6]
```diff
@@ -175,46 +228,129 @@ describe("Comments", () => {
         // but on `localhost`, cypress is fast enough
         // to execute the commands + assertion
         // before the redirect is done
-        openCommentsDrawer()
-        cy.get(COMMENTS_INPUT_SELECTOR).type(MOCK_COMMENT).blur()
-        submitComment()
```
Comment (seaerchin) [inline cypress/e2e/comments.spec.ts:180]: this is safe to remove cos it's rebased off the fix branch right?

File: cypress/e2e/comments.spec.ts [hunk 7]
```diff
@@ -153,8 +208,6 @@ describe("Comments", () => {
     describe("not a site member", () => {
       before(() => {
         removeOtherCollaborators()
-      })
-      beforeEach(() => {
```
Comment (seaerchin) [inline cypress/e2e/comments.spec.ts:157]: why remove this? if we do some permission changes in the test, this change makes it so that the permission changes will propagate to other tests in the same suite

File: cypress/e2e/comments.spec.ts [hunk 8]
```diff
@@ -175,46 +228,129 @@ describe("Comments", () => {
...
+        // as the backend recognises that we lack sufficient permissions
+        // in order to view this site
+        ignoreAuthError()
+        ignoreNotFoundError()
+        // NOTE: There is a redirect being done
+        // but on `localhost`, cypress is fast enough
+        // to execute the commands + assertion
+        // before the redirect is done
+
+        // // Assert
```
Comment (seaerchin) [inline cypress/e2e/comments.spec.ts:257]: nit: remove double comment

File: cypress/e2e/comments.spec.ts [hunk 9]
```diff
@@ -175,46 +228,129 @@ describe("Comments", () => {
...
       })
-      it("should not be able to create comments for closed review requests", () => {
+      it("should not be able to see/create comments for closed review requests", () => {
         // Arrange
         // Act
+        cy.visit(
+          `${CMS_BASEURL}/sites/${E2E_EMAIL_TEST_SITE.repo}/review/${reviewId}`
+        )
         // Assert
+        checkCommentDisabled()
```
Comment (seaerchin) [inline cypress/e2e/comments.spec.ts:335]: wrong assumption - this only checks on initial load if the **button** is disabled. we need to also assert that the button is disabled _after_ the user types something into the input. (if the assertion in `checkCommentDisabled` was instead changed to be both button + input then this would be ok)

File: cypress/e2e/comments.spec.ts [hunk 10]
```diff
@@ -175,46 +228,129 @@ describe("Comments", () => {
...
+      it("should not be able to see/create comments for non-existent review requests", () => {
         // Arrange
+        visitE2eEmailTestRepo()
+
         // Act
+
         // Assert
+        cy.contains(
+          `a[href^="/sites/${E2E_EMAIL_TEST_SITE.repo}/review/"]`
+        ).should("not.exist")
```
Comment (seaerchin) [inline cypress/e2e/comments.spec.ts:353]: at the point of testing this step, are we not still at the e2e test repo (url is `/dashboard`) and not the actual `/review/<id>` url?

File: src/layouts/ReviewRequest/Dashboard.tsx
Comment (seaerchin) [inline src/layouts/ReviewRequest/Dashboard.tsx:1]: should rebase off

**Comment by jandonyada** [comment]
Brother @caando , pls resolve the PR comments so we can close this chapter and move on with life 👼

File: cypress/e2e/comments.spec.ts (hunk 10 above)
Comment (caando) [reply cypress/e2e/comments.spec.ts:353]: For this test there is no review requests, so it goes to the site dashboard, then checks if there is a link to the currently opened review request. This exerts that the link does not exist.

File: cypress/e2e/comments.spec.ts (hunk 9 above)
Comment (caando) [reply cypress/e2e/comments.spec.ts:335]: For this test the user is unable to open the comments drawer to type stuff into the input box. Do you think I should skip this step and force the comment box to open?

File: cypress/e2e/comments.spec.ts (hunk 4 above)
Comment (caando) [reply cypress/e2e/comments.spec.ts:41]: Doesn't affect tests because most of them visit either this page or other pages on set up. I can add it back

File: cypress/e2e/comments.spec.ts (hunk 7 above)
Comment (caando) [reply cypress/e2e/comments.spec.ts:157]: The second test for `not a site member` does `setUserAsUnauthorised()` twice. Moving it to test set up instead.

File: cypress/e2e/comments.spec.ts (hunk 6 above)
Comment (caando) [reply cypress/e2e/comments.spec.ts:180]: This is passing test thou, users are supposed to be able to submit comments

File: cypress/api/reviewRequest.api.ts [hunk 11]
```diff
@@ -30,11 +30,29 @@ export const listReviewRequests = (): Cypress.Chainable<{ id: number }[]> => {
...
     reviewRequests.forEach(({ id }) => {
-      cy.request("DELETE", `${BASE_URL}/${id}`)
+      closeReviewRequest(id)
     })
   })
 }
 
-export const approveReviewRequest = (id: number): void => {
-  cy.request("POST", `${BASE_URL}/${id}/approve`)
+function removeResponse(response: any) {
```
Comment (seaerchin) [inline cypress/api/reviewRequest.api.ts:38]: why do we need this function?

File: cypress/e2e/comments.spec.ts (hunk 9 above)
Comment (seaerchin) [reply cypress/e2e/comments.spec.ts:340]: oh yea we updated the behaviour - ignore the above comment! (don't force open)

File: cypress/e2e/comments.spec.ts (hunk 10 above)
Comment (seaerchin) [reply cypress/e2e/comments.spec.ts:358]: ah ok this sounds logical - could we add this description above?

**Review by seaerchin — APPROVED** [review]
lgtm. please add the description of the link!
